src/cargo/util/errors.rs

```diff
@@ -11,141 +10,109 @@ use core::TargetKind;
-    fn chain_error<E, F>(self, callback: F) -> CargoResult<T>
-                         where E: CargoError, F: FnOnce() -> E;
-}
-
-#[derive(Debug)]
-struct ChainedError<E> {
-    error: E,
-    cause: Box<CargoError>,
-}
-
-impl<'a, T, F> ChainError<T> for F where F: FnOnce() -> CargoResult<T> {
-    fn chain_error<E, C>(self, callback: C) -> CargoResult<T>
-                         where E: CargoError, C: FnOnce() -> E {
-        self().chain_error(callback)
+error_chain! {
+    types {
+        CargoError, CargoErrorKind, CargoResultExt, CargoResult;
     }
-}
 
-impl<T, E: CargoError + 'static> ChainError<T> for Result<T, E> {
-    fn chain_error<E2: 'static, C>(self, callback: C) -> CargoResult<T>
-                         where E2: CargoError, C: FnOnce() -> E2 {
-        self.map_err(move |err| {
-            Box::new(ChainedError {
-                error: callback(),
-                cause: Box::new(err),
-            }) as Box<CargoError>
-        })
+    links {
+        CrateRegistry(registry::Error, registry::ErrorKind);
     }
-}
 
-impl<T> ChainError<T> for Box<CargoError> {
-    fn chain_error<E2, C>(self, callback: C) -> CargoResult<T>
-                         where E2: CargoError, C: FnOnce() -> E2 {
-        Err(Box::new(ChainedError {
-            error: callback(),
-            cause: self,
-        }))
+    foreign_links {
+        ParseSemver(semver::ReqParseError);
+        Semver(semver::SemVerError);
+        Io(io::Error);
+        SerdeJson(serde_json::Error);
+        TomlSer(toml::ser::Error);
+        TomlDe(toml::de::Error);
+        Term(term::Error);
+        ParseInt(num::ParseIntError);
+        ParseBool(str::ParseBoolError);
+        Parse(string::ParseError);
+        Git(git2::Error);
+        Curl(curl::Error);
     }
-}
 
-impl<T> ChainError<T> for Option<T> {
-    fn chain_error<E: 'static, C>(self, callback: C) -> CargoResult<T>
-                         where E: CargoError, C: FnOnce() -> E {
-        match self {
-            Some(t) => Ok(t),
-            None => Err(Box::new(callback())),
+    errors {
+        Internal(description: String){
+            description(&description)
+            display("{}", &description)
+        }
+        ProcessErrorKind(proc_err: ProcessError) {
+            description(&proc_err.desc)
+            display("{}", &proc_err.desc)
+        }
+        CargoTestErrorKind(test_err: CargoTestError) {
+            description(&test_err.desc)
+            display("{}", &test_err.desc)
+        }
+        HttpNot200(code: u32, url: String) {
+            description("failed to get a 200 response")
+            display("failed to get 200 response from `{}`, got {}", url, code)
         }
     }
 }
 
-impl<E: Error> Error for ChainedError<E> {
-    fn description(&self) -> &str { self.error.description() }
-}
+impl CargoError {
+    pub fn to_internal(self) -> Self {
+        //This is actually bad, because it loses the cause information for foreign_link
```

Comment:
This one was a bit messy. I can preserve `self.0`... but a foreign link error doesn't have anything in there, the auto impl just delegates to the underlying `Error::cause`, so the information gets lost anyway (and I couldn't come up with a way to turn `&Error` into a box, which is what would be needed to save the original cause. I think all uses of this wrapped foreign links rather than something higher-up the chain, but it's worth verifying it.